### DISCUSSION TITLE
Semantic Segmentation using PointNet/PPFNet on own data

### DISCUSSION THREAD

**Discussion by WizKhalista** [pyg-team/pytorch_geometric#4374] [Q&A]
Hi everyone,

i recently already had a question and I am not really progressing at the momemt, so i hope someone can help me.
### Task:
My main task is semantic segmentation of a scene, more detailed i have many objects recorded by a camera and i want to segment the Pointcloud, coming from the Camera so that every Object is assigned the correct class.
The Objects are Bolt/Nut/Cylinder/Quader as shown here:
![nut_pcl](https://user-images.githubusercontent.com/60539662/160636310-a12cfb6a-8ad5-45a9-83be-d766fd071eb3.png)![cylinder_pcl](https://user-images.githubusercontent.com/60539662/160635752-589ed4ce-5fa0-49dc-822b-8c648c9f6b0d.png)![bolt_pcl](https://user-images.githubusercontent.com/60539662/160636091-f7505ded-ace2-4dc1-86c0-019c6e939444.png)![quader_pcl](https://user-images.githubusercontent.com/60539662/160636359-6ded07ac-5a77-49e8-8ae7-1d307f80020b.png)

My Goal would be in a scene like this, to assign each object the correct class:
![scene_pcl](https://user-images.githubusercontent.com/60539662/160636619-aaadfe7e-4b5b-4804-82db-165e5317c47b.png)

### What i did:
I created a Dataset for the 4 Objects, where every Object is in different Rotations, since when predicting on a scene the orientation of the Objects varies. I split it into Train/Validation with 80/20 %.
Then i trained the Model and reached a Train accuracy of 98%-100% but the Validation accuracy is stagnating on 15%-25%. Since that was a sign of overfitting for me i already used Dropout etc. but i can get good results. Also when predicting multiple objects in one frame, every Point just gets the same class (as shown on the picture all Objects are Red=Nut).

Is my idea of training on single Objects (Bolt/Nut/Quader/Cylinder) and then expecting it to work on a scene with multiple Objects in it wrong?

I hope someone can help me and maybe already used the Pointnet/PPFNet for his own Segmentation purposes.
Thanks in advance!

Regards Marcel

**Comment by WizKhalista** [comment]
I think i figured it out, i just needed to use the PointConv Layer from the Pointnet instead of the PPFConv Layers and also there was an issue with building my Data Object from the Dataset. So i can see that the Model is clearly learning things.
Just one question left, when i train my Model with a single labeled pointcloud of an object at a time, will it be able to make a prediction for a whole scene with many objects of different classes in it?
### Pointclouds for training would look like this:
![nut](https://user-images.githubusercontent.com/60539662/160799836-e5ca4189-9b12-497b-9aee-12cfa9fa06d6.png)![bolt](https://user-images.githubusercontent.com/60539662/160799931-d3ac88f6-ae18-489a-9c58-e7f56f58de62.png)![quader](https://user-images.githubusercontent.com/60539662/160800020-34b5b0a9-b85a-43de-aca2-e9303ea2a0b7.png)![cylinder](https://user-images.githubusercontent.com/60539662/160800269-f58e446d-24dc-4c9b-9617-cb99ef58c7a5.png)

### The Scene i want to predict would look something like this:
![scene](https://user-images.githubusercontent.com/60539662/160800590-c33b42c0-0437-45cb-beb7-71c089fbca17.png)

Do you think that this would be possible?
Thank you in advance!

Best regards Marcel

**Reply by rusty1s** [reply]
I don't think this is possible without modification of your inference pipeline. One trivial approach would be to utilize a sliding window over your evaluation point cloud and try to predict the object in each window.

**Reply by WizKhalista** [reply]
Thank you @rusty1s  for your answer, that definitely makes sense.
The sliding window approach would be a good idea for scenes with not that many objects in it i think. My main goal is to make a prediction for such a scene:
![sceneToPredict](https://user-images.githubusercontent.com/60539662/161438250-0211022b-3d3b-4a93-835b-02d74ab8d4e7.png)
where the resulting pointcloud would look like this:
![pointcloudToPredict](https://user-images.githubusercontent.com/60539662/161438292-95acf226-00b3-4de5-aa9c-171aab4e8d6c.png)
Since there are many objects at different positions and sometimes occluding each other i think i need another approach.
I have read a few papers about how others are creating their Datasets for such a problem and i think i need to generate many scenes with those objects in different configurations in it and then train the Model. 
Since the Labeling of the pointclouds will be taking a lot of time i hope i don´t need too many scenes for it to learn.. :smile:

**Reply by rusty1s** [reply]
Yes, that works as well. In that case, this problem is interpreted as a segmentation problem rather than a classification problem.
